twine/settings.py

```diff
@@ -127,9 +128,22 @@ def __init__(
         self._handle_package_signing(
             sign=sign, sign_with=sign_with, identity=identity,
         )
-        # The following two rely on the parsed repository config
+        # _handle_certificates relies on the parsed repository config
         self._handle_certificates(cacert, client_cert)
-        self._handle_authentication(username, password, non_interactive)
+        auth_class = auth.Private if non_interactive else auth.Resolver
+        self.auth = auth_class(self, username, password)
```

Comment:
If you take all of my suggestions, I believe this would become something like:

```
self.auth = auth.Resolver(config, system, username, password, non_interactive)
```

Reply:
The latest implementation constructs the `auth` object in a single expression, keeping the details about Resolver selection encapsulated in the `auth` module and reduces the number of parameters to the `Resolver` constructor to just two. I believe this technique addresses some of the concern posed in this and other suggestions while still avoiding passing boolean parameters alongside a good deal of other paremeters.

```python
        self.auth = auth.Resolver.choose(not non_interactive)(
            self.repository_config,
            auth.CredentialInput(username, password),
        )
```

I'm pretty happy with this latest approach and I hope you are too.

Reply:
I'm going to proceed with this change so we can wrap up a release, but I'm happy to review incremental changes against this change.